lambdas/process-cri-callback/src/main/java/uk/gov/di/ipv/core/processcricallback/service/CriCheckingService.java

```diff
@@ -0,0 +1,249 @@
+
+        auditService.sendAuditEvent(
+                new AuditEvent(
+                        AuditEventTypes.IPV_CRI_AUTH_RESPONSE_RECEIVED,
+                        configService.getSsmParameter(ConfigurationVariable.COMPONENT_ID),
+                        auditEventUser,
+                        extensions));
+
+        if (!ALLOWED_OAUTH_ERROR_CODES.contains(error)) {
+            LOGGER.warn("Unknown Oauth error code received");
+        }
+
+        var visitedCriDetails = new VisitedCredentialIssuerDetailsDto(criId, null, false, error);
+        ipvSessionItem.addVisitedCredentialIssuerDetails(visitedCriDetails);
+
+        LogHelper.logCriOauthError("OAuth error received from CRI", error, errorDescription, criId);
+
+        return (switch (error) {
+            case OAuth2Error.ACCESS_DENIED_CODE -> JOURNEY_ACCESS_DENIED;
+            case OAuth2Error.TEMPORARILY_UNAVAILABLE_CODE -> JOURNEY_TEMPORARILY_UNAVAILABLE;
+            default -> JOURNEY_ERROR;
+        });
+    }
+
+    public void validateSessionIds(CriCallbackRequest callbackRequest)
+            throws InvalidCriCallbackRequestException {
+        var ipvSessionId = callbackRequest.getIpvSessionId();
+        var criOAuthSessionId = callbackRequest.getState();
+
+        if (StringUtils.isBlank(ipvSessionId)) {
+            if (!StringUtils.isBlank(criOAuthSessionId)) {
+                throw new InvalidCriCallbackRequestException(
+                        ErrorResponse.NO_IPV_FOR_CRI_OAUTH_SESSION);
+            }
+            throw new InvalidCriCallbackRequestException(ErrorResponse.MISSING_OAUTH_STATE);
+        }
+    }
+
+    public void validateCallbackRequest(
+            CriCallbackRequest callbackRequest, CriOAuthSessionItem criOAuthSessionItem)
+            throws InvalidCriCallbackRequestException {
+        var ipvSessionId = callbackRequest.getIpvSessionId();
+        var criId = callbackRequest.getCredentialIssuerId();
+        var state = callbackRequest.getState();
+        var authorisationCode = callbackRequest.getAuthorizationCode();
+
+        if (StringUtils.isBlank(authorisationCode)) {
+            throw new InvalidCriCallbackRequestException(ErrorResponse.MISSING_AUTHORIZATION_CODE);
+        }
+        if (StringUtils.isBlank(criId)) {
+            throw new InvalidCriCallbackRequestException(
+                    ErrorResponse.MISSING_CREDENTIAL_ISSUER_ID);
+        }
+        if (StringUtils.isBlank(ipvSessionId)) {
+            throw new InvalidCriCallbackRequestException(ErrorResponse.MISSING_IPV_SESSION_ID);
+        }
+        if (StringUtils.isBlank(state)) {
+            throw new InvalidCriCallbackRequestException(ErrorResponse.MISSING_OAUTH_STATE);
+        }
+        var persistedOauthState = getPersistedOauthState(criOAuthSessionItem);
+        if (!callbackRequest
+                .getState()
+                .equals(persistedOauthState)) { // might be ipvSessionItem.getCriOAuthSessionId() !=
+            // null ? ... : null
```

Comment:
Will it ever be valid for this to be be null? I wonder if we just make that explicit:
```
if (criOauthSessionItem == null || !state.equals(criOauthSessionItem.getCriOAuthSessionId()) {
  throw new InvalidCriCallbackRequestException(ErrorResponse.INVALID_OAUTH_STATE);
}
```